node/bin/src/command_source.rs

```diff
@@ -80,15 +83,31 @@ impl PipelineComponent for ExternalNodeCommandSource {
         output: mpsc::Sender<BlockCommand>,
     ) -> anyhow::Result<()> {
         // TODO: no need for a Stream in `replay_receiver` - just send to channel right away instead
-        let mut stream = replay_receiver(self.starting_block, self.replay_download_address.clone())
-            .await
-            .map_err(|err| {
-                tracing::error!(?err, "Failed to connect to main node to receive blocks");
-                err
-            })?;
+        let mut stream = replay_receiver(
+            self.starting_block,
+            self.record_overrides,
+            self.replay_download_address.clone(),
+        )
+        .await
+        .map_err(|err| {
+            tracing::error!(?err, "Failed to connect to main node to receive blocks");
+            err
+        })?;
 
         while let Some(command) = stream.next().await {
             tracing::debug!(?command, "Received block command from main node");
+
+            if let Some(up_to_block) = self.up_to_block
+                && command.block_number() > up_to_block
+            {
+                tracing::info!(
+                    up_to_block,
+                    "Reached up_to_block, halting external command source"
+                );
+                // Sleeping indefinitely.
+                futures::future::pending::<()>().await;
```

Comment:
Technically this should wait for `stop_receiver` but we have this problem in a lot of places. If it's difficult to get it from here feel free to ignore (but I'd at least leave a comment here)